plugins/CoreHome/vue/src/NumberFormatter/NumberFormatter.ts

```diff
@@ -0,0 +1,256 @@
+    // have a decimal point, and $valueParts[1] won't be set.
+    let minorDigits = valueParts[1] || '';
+
+    const usesGrouping = (pattern.indexOf(',') !== -1);
+
+    // if pattern has number groups, parse them.
+    if (usesGrouping) {
+      const primaryGroupMatches = pattern.match(/#+0/);
+      const primaryGroupSize = primaryGroupMatches?.[0].length || 0;
+      let secondaryGroupSize = primaryGroupMatches?.[0].length || 0;
+      const numberGroups = pattern.split(',');
+
+      // check for distinct secondary group size.
+      if (numberGroups.length > 2) {
+        secondaryGroupSize = numberGroups[1].length;
+      }
+
+      // Reverse the major digits, since they are grouped from the right.
+      const digits = majorDigits.split('').reverse();
+      // Group the major digits.
+      let groups = [];
+
+      groups.push(digits.splice(0, primaryGroupSize).reverse().join(''));
+
+      while (digits.length) {
+        groups.push(digits.splice(0, secondaryGroupSize).reverse().join(''));
+      }
+
+      // Reverse the groups and the digits inside of them.
+      groups = groups.reverse();
+      // Reconstruct the major digits.
+      majorDigits = groups.join(',');
+    }
+
+    if (minFractionDigits > 0) {
+      // Strip any trailing zeroes.
+      minorDigits = minorDigits.replace(/0+$/, '');
+      if (
+        minorDigits.length < minFractionDigits
+        && minorDigits.length < maxFractionDigits
+      ) {
+        // Now there are too few digits, re-add trailing zeroes
+        // until the desired length is reached.
+        const neededZeroes = minFractionDigits - minorDigits.length;
+        minorDigits += (new Array(neededZeroes + 1)).join('0');
+      }
+    }
+
+    // Assemble the final number and insert it into the pattern.
+    let result = minorDigits ? `${majorDigits}.${minorDigits}` : majorDigits;
+    result = pattern.replace(/#(?:[.,]#+)*0(?:[,.][0#]+)*/, result);
+
+    // Localize the number.
+    return this.replaceSymbols(result);
+  }
+
+  /**
+   * Replaces the placeholders with real symbols
+   *
+   * @param value
+   * @returns {string}
+   */
+  private replaceSymbols(value: string): string {
+    const replacements = {
+      '.': Matomo.numbers.symbolDecimal,
+      ',': Matomo.numbers.symbolGroup,
+      '+': Matomo.numbers.symbolPlus,
+      '-': Matomo.numbers.symbolMinus,
+      '%': Matomo.numbers.symbolPercent,
+    };
+
+    let newValue = '';
+    const valueParts = value.split('');
+
+    valueParts.forEach((val) => {
+      let valueReplaced = val;
+
+      Object.entries(replacements).some(([char, replacement]) => {
+        if (valueReplaced.indexOf(char) !== -1) {
+          valueReplaced = valueReplaced.replace(char, replacement);
+          return true;
+        }
+
+        return false;
+      });
+
+      newValue += valueReplaced;
+    });
+
+    return newValue;
+  }
+
+  private valOrDefault(def: number, val?: number): number {
```

Comment:
```suggestion
  private valOrDefault(val: number|undefined, def: number): number {
```

Being able to call the method without a value parameters makes not much sense here, and the parameter order is a bit confusing right now.